flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperRunningJobsRegistry.java

```diff
@@ -127,7 +129,14 @@ private String createZkPath(JobID jobID) {
     private void writeEnumToZooKeeper(JobID jobID, JobSchedulingStatus status) throws Exception {
         LOG.debug("Setting scheduling state for job {} to {}.", jobID, status);
         final String zkPath = createZkPath(jobID);
-        this.client.newNamespaceAwareEnsurePath(zkPath).ensure(client.getZookeeperClient());
-        this.client.setData().forPath(zkPath, status.name().getBytes(ENCODING));
+        final Stat stat = this.client.checkExists().forPath(zkPath);
+        if (stat != null) {
+            this.client.setData().forPath(zkPath, status.name().getBytes(ENCODING));
+        } else {
+            this.client
+                    .create()
+                    .creatingParentContainersIfNeeded()
+                    .forPath(zkPath, status.name().getBytes(ENCODING));
+        }
```

Comment:
In order to make this method rock solid, I think we need to wrap it in a while loop and retry the operation if something has changed between `this.client.checkExists().forPath(zkPath)` and one of the if branches. For example, it could have happened, that after the `checkExists` which returns `true`, that somebody else removes the node. Then the `this.client.setData().forPath(zkPath, status.name().getBytes(ENCODING));` should throw a `KeeperException.NoNodeException`. This should in practice not really happen but at the moment the creation/update procedure is not an atomic operation.

Reply:
The same can happen when calling `clearJob`.

Reply:
The current test is already testing with multiple paths.  The `setJobRunning` happens when there is no node and `setJobFinished` happens when the is a node.  `clearJob` is called on an existing node.  I added comments to this effect.  I added another check for `clearJob` when there is no node.  Also I added the loop as requested.